### PR TITLE
added some comment capabilities

### PR DESCRIPTION
**Pull request by flying-sheep**
last test fails because all the program statements are shuffeled.

i can’t find out why. can you look into it please?

### CODE REVIEW

**Comment by juliankrispel** [comment]
Awesome! I have some time over the weekend. If I don't get to look at it today I'll do it then! <3

**Comment by flying-sheep** [comment]
OK!

File: src/parser.js [hunk 1]
```diff
@@ -428,14 +428,71 @@ function mapBlockStatements(node, meta) {
...
+
+  const nearest = comments.map(comm => findNearest(comm, statements));
+  // attach those comments to those nodes where we found a line overlap
+  nearest.forEach((node, index) => {
+    if (!node) return;
+    const comm = comments[index];
+    const separateLines = fl(comm) !== fl(node) && ll(comm) !== ll(node);
+    // should be disjunct anyway: fc(a) < lc(a) < fc(b) < lc(b)
+    const leading = separateLines ? fl(comm) < fl(node) : lc(comm) < fc(node);
+    node.comments = [leading ? mapComment(comm, true, false) : mapComment(comm, false, true)];
```
Comment (juliankrispel) [inline src/parser.js]: this is the offending line, it's what leads the parser to throw out the line order...

**Comment by juliankrispel** [comment]
When I log all statement nodes I can still see some `locationData` or `loc` attributes defined in some of the nodes.

**Comment by juliankrispel** [comment]
I'm guessing that because some nodes have location info and some not the nodes are printed out of order...

File: src/parser.js (hunk 1 above)
Comment (flying-sheep) [inline src/parser.js]: i don’t get it neither…

**Comment by flying-sheep** [comment]
> `locationData` *or* `loc`

sure? i only see `loc`, and i guess they’re created by calling `recast.parse`.

**Comment by flying-sheep** [comment]
hmm, i tested if it changes anything to kill the loc in the end, but it doesn’t:

```js
function killLoc(node) {
  if (!node) return node;
  if ('loc' in node) delete node.loc;
  for (const key of Object.keys(node || {})) {
    if (typeof node[key] === 'object') {
      killLoc(node[key]);
    }
  }
  return node;
}
```

if you run this on the whole tree, nothing changes.

but maybe you could use it around some `recast.parse` call…

**Comment by juliankrispel** [comment]
@flying-sheep I took cues from your work and tried to figure this out in #19. I actually managed to simplify it a bit by attaching comments to the `EmptyStatement` type, this makes keeping track of location redundant. 

But the fundamental issue here is that recast doesn't print comments in order. I've come pretty close to figuring out what's the issue, I'll report back tomorrow

**Comment by juliankrispel** [comment]
@flying-sheep ok I got it. Basically, what [`printComments`](https://github.com/benjamn/recast/blob/master/lib/comments.js#L322) does is it prints all `leading` comments at the beginning of a block and `trailing` comments at the end of a block..

**Comment by flying-sheep** [comment]
ah, then i’ll close this and watch #19

**Pull request closed without merging**